blockfish-engine/src/ai/state.rs

```diff
@@ -72,7 +73,7 @@ impl State {
     /// Applies the given placement to this state, modifying the queue and matrix.
     pub fn place(&mut self, pl: &Place) {
         pl.shape.blit_to(&mut self.matrix, pl.tf);
-        self.reached_goal |= self.matrix.sift_rows();
+        self.reached_goal |= self.matrix.sift_rows(self.garbage_remaining);
```

Comment:
nit: maybe simpler to not modify sift_rows():
```suggestion
        let reached_bottom = self.matrix.sift_rows();
        self.reached_goal |= reached_bottom && self.garbage_remaining == 0;
```